contrib/smtp_proxy/filters/network/source/smtp_decoder.cc

```diff
@@ -0,0 +1,268 @@
+#include "contrib/smtp_proxy/filters/network/source/smtp_decoder.h"
+
+#include <vector>
+
+#include "absl/strings/str_split.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace NetworkFilters {
+namespace SmtpProxy {
+
+namespace {
+
+static const absl::string_view CRLF = "\r\n";
+
+// While https://www.rfc-editor.org/rfc/rfc5321.html#section-4.5.3.1.4
+// prescribes a max command line of 512 bytes, individual SMTP
+// extensions may increase this. e.g. extensions that add an
+// additional keyword to the MAIL FROM: command typically increase the
+// max mail command by the length of that keyword. For BINARYMIME
+// (rfc3030) this is 16 bytes, for SMTPUTF8 (rfc6531), it is 8
+// bytes. To the letter of the rfc, you need to understand all of the
+// extensions advertised by the server to determine the exact
+// limit. This is not a full implementation of SMTP; the limit here is
+// meant as a circuit breaker against obvious garbage. The real
+// implementation on either side is free to enforce a more precise
+// limit.
+const size_t kMaxCommandLine = 1024;
+// Likewise, the rfc reply limit is 512 bytes
+// https://www.rfc-editor.org/rfc/rfc5321.html#section-4.5.3.1.5
+const size_t kMaxReplyLine = 1024;
+
+// The smtp proxy filter only expects to consume the first ~3
+// responses from the upstream (greeting, ehlo, starttls), which
+// should be much smaller than this but again, don't buffer an
+// unbounded amount.
+const size_t kMaxReplyBytes = 65536;
+
+using RawLinePtr = std::unique_ptr<char[]>;
+} // anonymous namespace
+
+Decoder::Result getLine(Buffer::Instance& data, size_t start, RawLinePtr& raw_line,
+                        absl::string_view& line, size_t max_line) {
```

Comment:
Sorry, but I didn't get it. why `raw_line` and `line` are used? why not just use a normal `std::string&` here?

Reply:
1: I was trying to avoid doing data.toString() first in case the client sent us a bunch of garbage
2: The only way I could figure out how to copy n bytes (up to crlf) out of Buffer::Instance is to manually iterate data.getRawSlices(). Is there an easier way to do this?

Reply:
I rewrote this to copy directly from the buffer if it's contiguous in the first slice (which should be the common case) and slow-path through a temp otherwise.